pgmpy/estimators/NOTEARS.py

```diff
@@ -0,0 +1,420 @@
+            the loss function to be used in objective function
+
+        max_iter: int
+            maximum number of dual ascent steps
+
+        h_tol: float
+            exit if |h(w_est)| <= htol
+
+        rho_max: float
+            exit if rho >= rho_max
+
+        w_threshold: float
+            drop edge if |weight| < w_threshold
+
+        c: float
+            progress rate in the range (0, 1)
+
+        Returns
+        -------
+        Estimated model: pgmpy.base.DAG
+            The estimated model with (local) minimum value of objective function.
+
+        Examples
+        --------
+        """
+
+        # Step 0: Initial checks and setup
+        if loss_type not in ("l2", "logistic", "poisson"):
+            raise ValueError(
+                f"loss_type must be one of: l2, logistic, or poisson. Got: {loss_type}"
+            )
+        if c <= 0 or c >= 1:
+            raise ValueError("c (progress rate) must be in the range (0, 1). Aborting")
+
+        if show_progress and config.SHOW_PROGRESS:
+            iteration = trange(int(max_iter))
+        else:
+            iteration = range(int(max_iter))
+
+        if expert_knowledge is None:
+            expert_knowledge = ExpertKnowledge()
+
+        expert_knowledge._orient_temporal_forbidden_edges(DAG(), only_edges=False)
+        n = self.data.shape[1]
+        nodes = self.data.columns.to_list()
+
+        if config.get_backend() == "numpy":
+            data = np.array(self.data.values, dtype=config.get_dtype())
+        else:
+            data = torch.tensor(
+                self.data.values, dtype=config.get_dtype(), device=config.get_device()
+            )
+
+        # Step 1: Initialize starting estimates
+        adjacency_matrix_est, rho, alpha, h = (
+            self.backend.zeros(2 * n * n),
+            1.0,
+            0.0,
+            self.backend.inf,
+        )  # double w_est into (w_pos, w_neg)
+
+        forbidden_mask = self.backend.zeros((n, n))
+        for u, v in expert_knowledge.forbidden_edges:
+            i = nodes.index(u)
+            j = nodes.index(v)
+            forbidden_mask[i][j] = 1
+
+        fixed_mask = self.backend.zeros((n, n))
+        for u, v in expert_knowledge.required_edges:
+            i = nodes.index(u)
+            j = nodes.index(v)
+            fixed_mask[i][j] = 1
+
+        bounds = [
+            (0, 0) if i == j else (0, None)
+            for _ in range(2)
+            for i in range(n)
+            for j in range(n)
+        ]
+
+        # Why are we normalizing only for l2 - stabilze by reducing value of squared loss
+        if loss_type == "l2":
+            data = data - self.backend.mean(data, axis=0, keepdims=True)
+
+        # Step 2: Iterate until max_iter iterations
+        adjacency_matrix_new, h_new = None, None
+        for _ in iteration:
+
+            # Step 2.(a): Solve for new values of W and h
+            while rho < rho_max:
+                sol = sopt.minimize(

```

Comment:
This still doesn't work for pytorch backend as scipy would convert all tensors to numpy. Will need to use pytorch.optim.LBFGS for pytorch backend or can add a new function in compat_fns.py file.